runtime/src/gpu/cuda/gpu-cuda.c

```diff
@@ -113,6 +115,16 @@ void chpl_gpu_impl_init() {
     cuDeviceGetAttribute(&deviceClockRates[i], CU_DEVICE_ATTRIBUTE_CLOCK_RATE, device);
 
     chpl_gpu_primary_ctx[i] = context;
+
+    // TODO can we refactor some of this to chpl-gpu to avoid duplication
```

Comment:
I suppose we could by wrapping our own abstracted version of `cuModuleGetGlobal`. And maybe that's the right thing to do, though I think if we got updating the AMD library down to just running `hipify` I'd rather not add another abstraction layer.

I suppose now I'm also wondering what this would look like for Intel; if there's no equivalent for `hipify` there than maybe abstracting these lower level driver calls (`cuCtxSetCurrent`, `cuModuleGetGlobal`, etc.) is the right thing to do.

Reply:
I am imagining adjusting the context can be handled by the "impl" function we add for it.

One thing that I was thinking while I was working on this, was whether we need `chpl_gpu_impl_set_global("chpl_nodeID", chpl_nodeID);` instead. IOW, do we expose the addresses of globals in the GPU binary outside of the implementation layer, which "get global" would imply? While I don't see the harm of it, it feels redundant if all we'll ever do with them is to set them using other "impl" functions.

A bit too philosophical probably. If I had to decide right now, I might go with "set global" version. If you have any arguments for one or the other, I can easily be swayed.

Reply:
> IOW, do we expose the addresses of globals in the GPU binary outside of the implementation layer, which "get global" would imply? While I don't see the harm of it, it feels redundant if all we'll ever do with them is to set them using other "impl" functions.

I don't see a harm in it, we expose device pointers elsewhere right?

If having it as a higher level `chpl_gpu_impl_set_global` fits your design sensibilities better I'd say go with that.